File: src/main/scala/org/joisen/chapter05/TransformFilterTest.scala
package org.joisen.chapter05

import org.apache.flink.api.common.functions.{FilterFunction, MapFunction}
import org.apache.flink.streaming.api.scala._

/**
 * @Author Joisen
 * @Date 2022/11/20 15:52
 * @Version 1.0
 */
object TransformFilterTest {
  def main(args: Array[String]): Unit = {

    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream: DataStream[Event] = env.fromElements(Event("Mary", "./home", 1000L),
      Event("Bob", "./cart", 2000L))

    // 过滤用户Mary的所有点击事件
    // 方式1：使用 lambda 表达式
    stream.filter(_.user == "Mary").print("1")

    // 方式2：实现FilterFunction
    stream.filter(new UserFilter)

    env.execute()
  }

  class UserFilter extends FilterFunction[Event]{
    override def filter(t: Event): Boolean = t.user == "Bob"
  }

}

File: src/main/java/org/joisen/java/chapter05/TransformFilterTest.java
package org.joisen.java.chapter05;

import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * @Author Joisen
 * @Date 2022/12/5 20:48
 * @Version 1.0
 */
public class TransformFilterTest {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStreamSource<Event> stream = env.fromElements(new Event("Alice", "./home", 1000L),
                new Event("Mary", "./me", 2000L),
                new Event("Bob", "./cart", 3000L),
                new Event("Cary", "./prod?id=1", 4000L));

//        stream.filter(new MyFilter()).print();
        // 传入匿名类实现
        stream.filter(new FilterFunction<Event>() {
            @Override
            public boolean filter(Event event) throws Exception {
                return event.user.equals("Bob");
            }
        }).print();
        // λ表达式实现
        stream.filter(data -> data.user.equals("Alice")).print("λ ");



        env.execute();
    }

    public static class MyFilter implements FilterFunction<Event> {
        @Override
        public boolean filter(Event event) throws Exception {
            return event.user.equals("Mary");
        }
    }

}
